src/transformers/models/codegen/modeling_codegen.py

```diff
@@ -457,11 +476,12 @@ def forward(
         if token_type_ids is not None:
             token_type_ids = token_type_ids.view(-1, input_shape[-1])
 
-        if past_key_values is None:
-            past_length = 0
-            past_key_values = tuple([None] * len(self.h))
-        else:
-            past_length = past_key_values[0][0].size(-2)
+        past_length = 0
+        if use_cache:
+            use_legacy_cache = not isinstance(past_key_values, Cache)
+            if use_legacy_cache:
+                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+            past_length = past_key_values.get_seq_length()
```

Comment:
this is super ugly IMO. 
I would really love for us to just no add this code that ended up being copy pasted everywhere because we were lazy to update when porting new models. 

Let's use cache positions. And let's already deprecate the casting and legacy cache etc, to make sure we only do this for one revision!

Reply:
I see, I wanted to add `cache_position` as part of StaticCache support but it can be added in this PR also. Wondering about deprecation cycle for Cache class, afaik there were no deprecation warning about that before so we would still have to keep the ugly hack and add a warning message? 

I'm pro of totally getting rid of old cache, if it doesn't break BC

Reply:
Ah I see that Llama on main had its preparation modified already assuming that the inputs is always Cache object. Oke, then it makes sense to get rid of legacy_cache in forward also

Reply:
Yeah it was not done before but this PR IMO is a good time to do it! So keep this code, but add a warning saying, we are automatically converting your cache from tuple to dynamic cache class (SHould only be triggered outside generate because generate should already pass a cache class!)

Yep it makes sense but let's not be too brutal in case some people still use it! We give them one release until we totally remove it!

Reply:
Oke, added cache position and deprecation warnings to all models from this PR. I'll add the same deprecation warning to all models that already support cache class in another PR. This one is ready for review, tests are passing on my end!